bitfield/impl/src/bitfield.rs

```diff
@@ -0,0 +1,245 @@
+use proc_macro2::TokenStream as TokenStream2;
+use syn::{
+    self,
+    Token,
+    punctuated::Punctuated,
+    spanned::Spanned as _,
+    parse::{
+        Parse,
+        ParseStream,
+        Result as ParseResult,
+    }
+};
+use quote::{
+    quote,
+    quote_spanned,
+    ToTokens,
+};
+
+pub fn generate(args: TokenStream, input: TokenStream) -> TokenStream {
+    match bitfield_impl(args.into(), input.into()) {
+        Ok(output) => output.into(),
+        Err(err) => err.to_compile_error().into(),
+    }
+}
+
+fn bitfield_impl(args: TokenStream2, input: TokenStream2) -> Result<TokenStream2, syn::Error> {
+    let _ = args;
+    let input = syn::parse::<BitfieldStruct>(input.into())?;
+    input.validate()?;
+    Ok(input.into_token_stream())
+}
+
+struct BitfieldStruct {
+    ast: syn::ItemStruct,
+}
+
+/// Represents the `bitfield` specific attribute `#[bits = N]`.
+struct BitsAttributeArgs {
+    eq_tok: Token![=],
+    size: syn::LitInt,
+}
+
+impl syn::parse::Parse for BitsAttributeArgs {
+    fn parse(input: &syn::parse::ParseBuffer) -> syn::Result<Self> {
+        Ok(BitsAttributeArgs {
+            eq_tok: input.parse()?,
+            size: input.parse()?,
+        })
+    }
+}
+
+impl Parse for BitfieldStruct {
+    fn parse(input: ParseStream) -> ParseResult<Self> {
+        Ok(Self {
+            ast: input.parse()?,
+        })
+    }
+}
+
+impl ToTokens for BitfieldStruct {
+    fn to_tokens(&self, tokens: &mut TokenStream2) {
+
+        let mut size = Punctuated::<syn::ExprPath, Token![+]>::new();
+        if let syn::Fields::Named(fields_named) = &self.ast.fields {
+            for field in fields_named.named.iter() {
+                let ty = &field.ty;
+                size.push(syn::parse_quote!( <#ty as Specifier>::BITS ))
+            }
+        } else {
+            unreachable!()
+        }
+        let methods = self.methods_into_tokens();
+        for attr in &self.ast.attrs {
+            attr.to_tokens(tokens)
+        }
+        let ident = &self.ast.ident;
+        tokens.extend(quote!{
+            #[repr(C)]
+            pub struct #ident
+            {
+                data: [u8; (#size) / 8],
+            }
+
+            impl bitfield::checks::CheckTotalSizeMultipleOf8 for #ident {
+                type Size = bitfield::checks::TotalSize<[(); (#size) % 8]>;
+            }
+
+            impl #ident
+            {}
+
+            impl #ident
+            {
+                pub fn new() -> Self {
+                    Self {
+                        data: [0; (#size) / 8],
+                    }
+                }
+
+                #methods
+            }
+        });
+    }
+}
+
+impl BitfieldStruct {
+    fn methods_into_tokens(&self) -> TokenStream2 {
+        let mut tokens = quote! {};
+        self.methods_to_tokens(&mut tokens);
+        tokens
+    }
+
+    fn methods_to_tokens(&self, tokens: &mut TokenStream2) {
+        if let syn::Fields::Named(fields_named) = &self.ast.fields {
```

Comment:
I would recommend writing this with a `match` so that you avoid indenting the next 105 lines.

```rust
let fields = match &self.ast.fields {
    Fields::Named(fields) => fields,
    _ => unreachable!(),
};
```

Reply:
Indeed much more readable!